### PR TITLE
Added parser module

### PR DESCRIPTION
**Pull request by Saif807380**
### Added Parser Module

Since almost all functions differentiate between categorical columns and regular columns, I added a new module after the Reader module to calculate the categorical columns and store their names in `params["cat_cols"]`

### Encoder

- Renamed `EncodeData` to `Encoder`
- Removed `DateTime` check until it can be fixed
- Updated condition to check for `int` or `str` column

### Reader

- Renamed `ReadData` to `Reader`
- Refactored `Reader` to support `params`
- Removed dataframe summary functions

### Base Pipeline

- Refactored base pipeline as a direct consequence of refactoring `Reader`

### Config

- Refactored `save_config` as a direct consequence of refactoring base pipeline

### Testing

- Added tests for `Parser`
- Updated tests for base pipeline, config, `Encoder` and `Reader`

### CODE REVIEW

**Review by Bhumika-Kothwal — CHANGES_REQUESTED** [review]

File: tests/test_parse.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    ord_dict1["Size"] = None
+    params = {
+        "train_df": train_csv,
+        "target_label": "Price",
+        "ord_dict": ord_dict1,
+    }
+    with pytest.raises(ValueError):
+        parser = Parser()
+        parser.parse_dataset(params=params)
+
```
Comment (Bhumika-Kothwal) [inline tests/test_parse.py:38]: Can we also add test for incorrect cat cols passed by user?

**Review by Bhumika-Kothwal — CHANGES_REQUESTED** [review]

File: preprocessy/parse/_summarize.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        rows = self.train_df.shape[0]
+        rows = 0.5 * rows
+        self.cat_cols = []
+        for col in self.train_df.columns:
+            if col not in self.ord_cols:
+                if self.train_df[col].dtype == "object" and (
+                    "$" in self.train_df[col][0]
+                    or self.train_df[col].str.contains(",").any()
+                ):
+                    print("money")
```
[the remark this answers is not in the document]
Comment (Saif807380) [reply preprocessy/parse/_summarize.py:57]: I'll remove that

File: tests/test_parse.py (hunk 1 above)
Comment (Saif807380) [reply tests/test_parse.py:38]: I'll add a new test

File: preprocessy/parse/_summarize.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    def parse_dataset(self, params):
+        if "train_df" in params.keys():
+            self.train_df = params["train_df"]
+        if "test_df" in params.keys():
+            self.test_df = params["test_df"]
+        if "cat_cols" in params.keys():
+            self.cat_cols = params["cat_cols"]
+            for col in self.cat_cols:
+                if col not in self.train_df or (
```
Comment (Bhumika-Kothwal) [inline preprocessy/parse/_summarize.py:79]: We will need to validate user give cat cols after calling validate function i.e. line 90 as train and test df need to be validated before using them

**Review by Bhumika-Kothwal — CHANGES_REQUESTED** [review]

**Review by Bhumika-Kothwal — APPROVED** [review]

**Pull request merged**